File: src/main/scala/com/snowflake/snowpark/Row.scala
package com.snowflake.snowpark

import java.sql.{Date, Time, Timestamp}
import com.snowflake.snowpark.internal.ErrorMessage
import com.snowflake.snowpark.types.{Geography, Geometry, Variant}

import scala.util.hashing.MurmurHash3

/**
 * @since 0.1.0
 */
object Row {

  /**
   * Returns a [[Row]] based on the given values.
   * @since 0.1.0
   */
  def apply(values: Any*): Row = new Row(values.toArray)

  /**
   * Return a [[Row]] based on the values in the given Seq.
   * @since 0.1.0
   */
  def fromSeq(values: Seq[Any]): Row = new Row(values.toArray)

  /**
   * Return a [[Row]] based on the values in the given Array.
   * @since 0.2.0
   */
  def fromArray(values: Array[Any]): Row = new Row(values)

  private[snowpark] def fromMap(map: Map[String, Any]): Row =
    new SnowflakeObject(map)
}

private[snowpark] class SnowflakeObject private[snowpark]
  (private[snowpark] val map: Map[String, Any]) extends Row(map.values.toArray) {
  override def toString: String = convertValueToString(this)
}

/**
 * Represents a row returned by the evaluation of a [[DataFrame]].
 *
 * @groupname getter Getter Functions
 * @groupname utl Utility Functions
 * @since 0.1.0
 */
class Row protected (values: Array[Any]) extends Serializable {

  /**
   * Converts this [[Row]] to a Seq
   * @since 0.1.0
   * @group utl
   */
  def toSeq: Seq[Any] = values.toSeq

  /**
   * Total number of [[Column]] in this [[Row]]. Alias of [[length]]
   * @group utl
   * @since 0.1.0
   */
  def size: Int = length

  /**
   * Total number of [[Column]] in this [[Row]]
   * @since 0.1.0
   * @group utl
   */
  def length: Int = values.length

  /**
   * Returns the value of the column in the row at the given index. Alias of [[get]]
   * @since 0.1.0
   * @group getter
   */
  def apply(index: Int): Any = get(index)

  /**
   * Returns the value of the column in the row at the given index.
   * @since 0.1.0
   * @group getter
   */
  def get(index: Int): Any = values(index)

  /**
   * Returns a clone of this row.
   * @since 0.1.0
   * @group utl
   */
  def copy(): Row = new Row(values)

  /**
   * Returns a clone of this row object. Alias of [[copy]]
   * @since 0.1.0
   * @group utl
   */
  override def clone(): AnyRef = copy()

  /**
   * Returns true iff the given row equals this row.
   * @since 0.1.0
   * @group utl
   */
  override def equals(obj: Any): Boolean =
    if (!obj.isInstanceOf[Row]) {
      false
    } else {
      val other = obj.asInstanceOf[Row]
      if (length != other.length) {
        false
      } else {
        (0 until length).forall { index =>
          (this(index), other(index)) match {
            case (d1: Double, d2: Double) if d1.isNaN && d2.isNaN => true
            case (v1, v2) => v1 == v2
          }
        }
      }
    }

  /**
   * Calculates hashcode of this row.
   * @since 0.1.0
   * @group utl
   */
  override def hashCode(): Int = {
    var n = 0
    var h = MurmurHash3.seqSeed
    val len = length
    while (n < len) {
      h = MurmurHash3.mix(h, apply(n).##)
      n += 1
    }
    MurmurHash3.finalizeHash(h, n)
  }

  /**
   * Returns true if the value of the column at the given index is null, otherwise, returns false.
   * @since 0.1.0
   * @group utl
   */
  def isNullAt(index: Int): Boolean = get(index) == null

  /**
   * Returns the value of the column at the given index as a Boolean value
   * @since 0.1.0
   * @group getter
   */
  def getBoolean(index: Int): Boolean = getAnyValAs[Boolean](index)

  /**
   * Returns the value of the column at the given index as a Byte value.
   * Casts Short, Int, Long number to Byte if possible.
   * @since 0.1.0
   * @group getter
   */
  def getByte(index: Int): Byte = get(index) match {
    case byte: Byte => byte
    case short: Short if short <= Byte.MaxValue && short >= Byte.MinValue => short.toByte
    case int: Int if int <= Byte.MaxValue && int >= Byte.MinValue => int.toByte
    case long: Long if long <= Byte.MaxValue && long >= Byte.MinValue => long.toByte
    case other =>
      throw ErrorMessage.MISC_CANNOT_CAST_VALUE(other.getClass.getName, s"$other", "Byte")
  }

  /**
   * Returns the value of the column at the given index as a Short value.
   * Casts Byte, Int, Long number to Short if possible.
   * @since 0.1.0
   * @group getter
   */
  def getShort(index: Int): Short = get(index) match {
    case byte: Byte => byte.toShort
    case short: Short => short
    case int: Int if int <= Short.MaxValue && int >= Short.MinValue => int.toShort
    case long: Long if long <= Short.MaxValue && long >= Short.MinValue => long.toShort
    case other =>
      throw ErrorMessage.MISC_CANNOT_CAST_VALUE(other.getClass.getName, s"$other", "Short")
  }

  /**
   * Returns the value of the column at the given index as a Int value.
   * Casts Byte, Short, Long number to Int if possible.
   * @since 0.1.0
   * @group getter
   */
  def getInt(index: Int): Int = get(index) match {
    case byte: Byte => byte.toInt
    case short: Short => short.toInt
    case int: Int => int
    case long: Long if long <= Int.MaxValue && long >= Int.MinValue => long.toInt
    case other =>
      throw ErrorMessage.MISC_CANNOT_CAST_VALUE(other.getClass.getName, s"$other", "Int")
  }

  /**
   * Returns the value of the column at the given index as a Long value.
   * Casts Byte, Short, Int number to Long if possible.
   * @since 0.1.0
   * @group getter
   */
  def getLong(index: Int): Long = get(index) match {
    case byte: Byte => byte.toLong
    case short: Short => short.toLong
    case int: Int => int.toLong
    case long: Long => long
    case other =>
      throw ErrorMessage.MISC_CANNOT_CAST_VALUE(other.getClass.getName, s"$other", "Long")
  }

  /**
   * Returns the value of the column at the given index as a Float value.
   * Casts Byte, Short, Int, Long and Double number to Float if possible.
   * @since 0.1.0
   * @group getter
   */
  def getFloat(index: Int): Float = get(index) match {
    case float: Float => float
    case double: Double if double <= Float.MaxValue && double >= Float.MinValue => double.toFloat
    case byte: Byte => byte.toFloat
    case short: Short => short.toFloat
    case int: Int => int.toFloat
    case long: Long => long.toFloat
    case other =>
      throw ErrorMessage.MISC_CANNOT_CAST_VALUE(other.getClass.getName, s"$other", "Float")
  }

  /**
   * Returns the value of the column at the given index as a Double value.
   * Casts Byte, Short, Int, Long, Float number to Double.
   * @since 0.1.0
   * @group getter
   */
  def getDouble(index: Int): Double = get(index) match {
    case float: Float => float.toDouble
    case double: Double => double
    case byte: Byte => byte.toDouble
    case short: Short => short.toDouble
    case int: Int => int.toDouble
    case long: Long => long.toDouble
    case other =>
      throw ErrorMessage.MISC_CANNOT_CAST_VALUE(other.getClass.getName, s"$other", "Double")
  }

  /**
   * Returns the value of the column at the given index as a String value.
   * Returns geography data as string, if geography data of GeoJSON, WKT or EWKT is found.
   * @since 0.1.0
   * @group getter
   */
  def getString(index: Int): String = {
    get(index) match {
      case variant: Variant => variant.toString
      case geo: Geography => geo.toString
      case geo: Geometry => geo.toString
      case array: Array[_] => new Variant(array).toString
      case seq: Seq[_] => new Variant(seq).toString
      case map: Map[_, _] => new Variant(map).toString
      case _ => getAs[String](index)
    }
  }

  /**
   * Returns the value of the column at the given index as a Byte array value.
   * @since 0.2.0
   * @group getter
   */
  def getBinary(index: Int): Array[Byte] = getAs[Array[Byte]](index)

  /**
   * Returns the value of the column at the given index as a BigDecimal value
   * @since 0.1.0
   * @group getter
   */
  def getDecimal(index: Int): java.math.BigDecimal = getAs[java.math.BigDecimal](index)

  /**
   * Returns the value of the column at the given index as a Date value
   * @since 0.1.0
   * @group getter
   */
  def getDate(index: Int): Date = getAs[Date](index)

  /**
   * Returns the value of the column at the given index as a Time value
   * @since 0.2.0
   * @group getter
   */
  def getTime(index: Int): Time = getAs[Time](index)

  /**
   * Returns the value of the column at the given index as a Timestamp value
   * @since 0.2.0
   * @group getter
   */
  def getTimestamp(index: Int): Timestamp = getAs[Timestamp](index)

  /**
   * Returns the value of the column at the given index as Variant class
   * @since 0.2.0
   * @group getter
   */
  def getVariant(index: Int): Variant = new Variant(getString(index))

  /**
   * Returns the value of the column at the given index as Geography class
   * @since 0.2.0
   * @group getter
   */
  def getGeography(index: Int): Geography = getAs[Geography](index)

  /**
   * Returns the value of the column at the given index as Geometry class
   *
   * @since 1.12.0
   * @group getter
   */
  def getGeometry(index: Int): Geometry = getAs[Geometry](index)

  /**
   * Returns the value of the column at the given index as a Seq of Variant
   * @since 0.2.0
   * @group getter
   */
  def getSeqOfVariant(index: Int): Seq[Variant] =
    new Variant(getString(index)).asSeq()

  /**
   * Returns the value of the column at the given index as a java map of Variant
   * @since 0.2.0
   * @group getter
   */
  def getMapOfVariant(index: Int): Map[String, Variant] =
    new Variant(getString(index)).asMap()

  protected def convertValueToString(value: Any): String =
    value match {
      case null => "null"
      case map: Map[_, _] =>
        map
          .map {
            case (key, value) => s"${convertValueToString(key)}:${convertValueToString(value)}"
          }
          .mkString("Map(", ",", ")")
      case binary: Array[Byte] => s"Binary(${binary.mkString(",")})"
      case strValue: String => s""""$strValue""""
      case arr: Array[_] =>
        arr.map(convertValueToString).mkString("Array(", ",", ")")
      case obj: SnowflakeObject =>
        obj.map.map {
          case (key, value) =>
            s"$key:${convertValueToString(value)}"
        }.mkString("Object(", ",", ")")
      case other => other.toString
    }

  /**
   * Returns a string value to represent the content of this row
   * @since 0.1.0
   * @group utl
   */
  override def toString: String =
    values
      .map(convertValueToString)
      .mkString("Row[", ",", "]")

  private def getAs[T](index: Int): T = get(index).asInstanceOf[T]

  private def getAnyValAs[T <: AnyVal](index: Int): T =
    if (isNullAt(index)) throw new NullPointerException(s"Value at index $index is null")
    else getAs[T](index)

}

File: src/main/java/com/snowflake/snowpark_java/Row.java
package com.snowflake.snowpark_java;

import com.snowflake.snowpark.internal.JavaUtils;
import com.snowflake.snowpark_java.types.Geography;
import com.snowflake.snowpark_java.types.Geometry;
import com.snowflake.snowpark_java.types.InternalUtils;
import com.snowflake.snowpark_java.types.Variant;
import java.io.Serializable;
import java.math.BigDecimal;
import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.util.*;

/**
 * Represents a row returned by the evaluation of a {@code DataFrame}.
 *
 * @see com.snowflake.snowpark_java.DataFrame DataFram
 * @since 0.9.0
 */
public class Row implements Serializable, Cloneable {
  private final com.snowflake.snowpark.Row scalaRow;

  // package private constructor, internal use only.
  Row(com.snowflake.snowpark.Row row) {
    this.scalaRow = row;
  }

  /**
   * Creates a {@code Row} based on the values in the given array.
   *
   * @param values {@code Row} elements
   * @since 0.9.0
   */
  public Row(Object[] values) {
    this(com.snowflake.snowpark.Row.fromArray(javaObjectToScalaObject(values)));
  }

  private static Object[] javaObjectToScalaObject(Object[] input) {
    Object[] result = Arrays.copyOf(input, input.length);
    for (int i = 0; i < result.length; i++) {
      if (result[i] instanceof Variant) {
        result[i] = InternalUtils.toScalaVariant((Variant) result[i]);
      } else if (result[i] instanceof Variant[]) {
        Variant[] javaVariantArray = (Variant[]) result[i];
        com.snowflake.snowpark.types.Variant[] resultArray =
            new com.snowflake.snowpark.types.Variant[javaVariantArray.length];
        for (int idx = 0; idx < resultArray.length; idx++) {
          resultArray[idx] = InternalUtils.toScalaVariant(javaVariantArray[idx]);
        }
        result[i] = resultArray;
      } else if (result[i] instanceof Map<?, ?>) {
        result[i] = JavaUtils.javaMapToScalaWithVariantConversion((Map<Object, Object>) result[i]);
      } else if (result[i] instanceof Geography) {
        result[i] =
            com.snowflake.snowpark.types.Geography.fromGeoJSON(((Geography) result[i]).asGeoJSON());
      } else if (result[i] instanceof Geometry) {
        result[i] = com.snowflake.snowpark.types.Geometry.fromGeoJSON(result[i].toString());
      }
    }
    return result;
  }

  /**
   * Converts this {@code Row} to a {@code List} of {@code Object}.
   *
   * @return A {@code List} contains {@code Row} elements.
   * @since 0.9.0
   */
  public List<Object> toList() {
    return JavaUtils.seqToList(scalaRow.toSeq());
  }

  /**
   * Counts the number of column in this {@code Row}.
   *
   * @return An integer number represents the size of this {@code Row}
   * @since 0.9.0
   */
  public int size() {
    return scalaRow.size();
  }

  /**
   * Creates a clone of this {@code Row} object.
   *
   * @return A new {@code Row} object containing same elements as this {@code Row}
   * @since 0.9.0
   */
  @Override
  public Row clone() {
    Row cloned;
    try {
      cloned = (Row) super.clone();
    } catch (CloneNotSupportedException e) {
      cloned = new Row(scalaRow);
    }
    return cloned;
  }

  /**
   * Verifies the equality of two {@code Row} objects.
   *
   * @param other The {@code Row} object to be compared to
   * @return true if the given row equals this row.
   * @since 0.9.0
   */
  @Override
  public boolean equals(Object other) {
    if (other instanceof Row) {
      Row otherRow = (Row) other;
      return this.scalaRow.equals(otherRow.scalaRow);
    }
    return false;
  }

  /**
   * Calculates the hash code of this Row object
   *
   * @return An int number representing the hash code
   * @since 0.9.0
   */
  @Override
  public int hashCode() {
    return scalaRow.hashCode();
  }

  /**
   * Retrieves the value of column in the {@code Row} at the given index.
   *
   * @param index The index of the target column
   * @return The value of the column at the given index
   */
  public Object get(int index) {
    return toJavaValue(scalaRow.get(index));
  }

  private static Object toJavaValue(Object value) {
    if (value instanceof com.snowflake.snowpark.types.Variant) {
      return InternalUtils.createVariant((com.snowflake.snowpark.types.Variant) value);
    } else if (value instanceof com.snowflake.snowpark.types.Geography) {
      return Geography.fromGeoJSON(((com.snowflake.snowpark.types.Geography) value).asGeoJSON());
    } else if (value instanceof com.snowflake.snowpark.types.Geometry) {
      return Geometry.fromGeoJSON(value.toString());
    } else if (value instanceof com.snowflake.snowpark.types.Variant[]) {
      com.snowflake.snowpark.types.Variant[] scalaVariantArray =
          (com.snowflake.snowpark.types.Variant[]) value;
      Variant[] resultArray = new Variant[scalaVariantArray.length];
      for (int idx = 0; idx < scalaVariantArray.length; idx++) {
        resultArray[idx] = InternalUtils.createVariant(scalaVariantArray[idx]);
      }
      return resultArray;
    } else if (value instanceof scala.collection.immutable.Map<?, ?>) {
      scala.collection.immutable.Map<?, ?> input = (scala.collection.immutable.Map<?, ?>) value;
      Map<Object, Object> result = new HashMap<>();
      // key is either Long or String, no need to convert values
      input.foreach(x -> result.put(x._1, toJavaValue(x._2)));
      return result;
    } else if (value instanceof Object[]) {
      Object[] arr = (Object[]) value;
      List<Object> result = new ArrayList<>(arr.length);
      for (Object x : arr) {
        result.add(toJavaValue(x));
      }
      return result;
    } else {
      return value;
    }
  }

  /**
   * Verifies if the value of the column at the given index is null.
   *
   * @param index The index of target column
   * @return true if the value is null at the given index
   * @since 0.9.0
   */
  public boolean isNullAt(int index) {
    return scalaRow.isNullAt(index);
  }

  /**
   * Retrieves the value of the column at the given index as a boolean value.
   *
   * @param index The index of target column
   * @return The boolean value of the column at the given index
   * @since 0.9.0
   */
  public boolean getBoolean(int index) {
    return scalaRow.getBoolean(index);
  }

  /**
   * Retrieves the value of the column at the given index as a byte value.
   *
   * @param index The index of target column
   * @return The byte value of the column at the given index
   * @since 0.9.0
   */
  public byte getByte(int index) {
    return scalaRow.getByte(index);
  }

  /**
   * Retrieves the value of the column at the given index as a short value.
   *
   * @param index The index of target column
   * @return The short value of the column at the given index
   * @since 0.9.0
   */
  public short getShort(int index) {
    return scalaRow.getShort(index);
  }

  /**
   * Retrieves the value of the column at the given index as a int value.
   *
   * @param index The index of target column
   * @return The int value of the column at the given index
   * @since 0.9.0
   */
  public int getInt(int index) {
    return scalaRow.getInt(index);
  }

  /**
   * Retrieves the value of the column at the given index as a long value.
   *
   * @param index The index of target column
   * @return The long value of the column at the given index
   * @since 0.9.0
   */
  public long getLong(int index) {
    return scalaRow.getLong(index);
  }

  /**
   * Retrieves the value of the column at the given index as a float value.
   *
   * @param index The index of target column
   * @return The float value of the column at the given index
   * @since 0.9.0
   */
  public float getFloat(int index) {
    return scalaRow.getFloat(index);
  }

  /**
   * Retrieves the value of the column at the given index as a double value.
   *
   * @param index The index of target column
   * @return The double value of the column at the given index
   * @since 0.9.0
   */
  public double getDouble(int index) {
    return scalaRow.getDouble(index);
  }

  /**
   * Retrieves the value of the column at the given index as a String value.
   *
   * @param index The index of target column
   * @return The String value of the column at the given index
   * @since 0.9.0
   */
  public String getString(int index) {
    return scalaRow.getString(index);
  }

  /**
   * Retrieves the value of the column at the given index as an array of byte.
   *
   * @param index The index of target column
   * @return An array of byte representing the binary value
   * @since 0.9.0
   */
  public byte[] getBinary(int index) {
    return scalaRow.getBinary(index);
  }

  /**
   * Retrieves the value of the column at the given index as a Date value.
   *
   * @param index The index of target column
   * @return The Date value of the column at the given index
   * @since 0.9.0
   */
  public Date getDate(int index) {
    return scalaRow.getDate(index);
  }

  /**
   * Retrieves the value of the column at the given index as a Time value.
   *
   * @param index The index of target column
   * @return The Time value of the column at the given index
   * @since 0.9.0
   */
  public Time getTime(int index) {
    return scalaRow.getTime(index);
  }

  /**
   * Retrieves the value of the column at the given index as a Timestamp value.
   *
   * @param index The index of target column
   * @return The Timestamp value of the column at the given index
   * @since 0.9.0
   */
  public Timestamp getTimestamp(int index) {
    return scalaRow.getTimestamp(index);
  }

  /**
   * Retrieves the value of the column at the given index as a BigDecimal value.
   *
   * @param index The index of target column
   * @return The BigDecimal value of the column at the given index
   * @since 0.9.0
   */
  public BigDecimal getDecimal(int index) {
    return scalaRow.getDecimal(index);
  }

  /**
   * Retrieves the value of the column at the given index as a Variant value.
   *
   * @param index The index of target column
   * @return The Variant value of the column at the given index
   * @since 0.9.0
   */
  public Variant getVariant(int index) {
    return InternalUtils.createVariant(scalaRow.getVariant(index));
  }

  /**
   * Retrieves the value of the column at the given index as a Geography value.
   *
   * @param index The index of target column
   * @return The Geography value of the column at the given index
   * @since 0.9.0
   */
  public Geography getGeography(int index) {
    return Geography.fromGeoJSON(scalaRow.getGeography(index).asGeoJSON());
  }

  /**
   * Retrieves the value of the column at the given index as a Geometry value.
   *
   * @param index The index of target column
   * @return The Geometry value of the column at the given index
   * @since 1.12.0
   */
  public Geometry getGeometry(int index) {
    return Geometry.fromGeoJSON(scalaRow.getGeometry(index).toString());
  }

  /**
   * Retrieves the value of the column at the given index as a list of Variant.
   *
   * @param index The index of target column
   * @return A list of Variant
   * @since 0.9.0
   */
  public List<Variant> getListOfVariant(int index) {
    scala.collection.Seq<com.snowflake.snowpark.types.Variant> scalaSeq =
        this.scalaRow.getSeqOfVariant(index);
    List<Variant> result = new ArrayList<>(scalaSeq.size());
    for (int i = 0; i < scalaSeq.size(); i++) {
      result.add(InternalUtils.createVariant(scalaSeq.apply(i)));
    }
    return result;
  }

  /**
   * Retrieves the value of the column at the given index as a map of Variant
   *
   * @param index The index of target column
   * @return A map from String to Variant
   * @since 0.9.0
   */
  public Map<String, Variant> getMapOfVariant(int index) {
    scala.collection.Map<String, com.snowflake.snowpark.types.Variant> scalaMap =
        this.scalaRow.getMapOfVariant(index);
    Map<String, Variant> result = new HashMap<>();
    scala.collection.Seq<String> keys = scalaMap.keys().toSeq();
    for (int i = 0; i < keys.size(); i++) {
      result.put(keys.apply(i), InternalUtils.createVariant(scalaMap.apply(keys.apply(i))));
    }
    return result;
  }

  /**
   * Retrieves the value of the column at the given index as a list of Object.
   *
   * @param index The index of target column
   * @return A list of Object
   * @since 1.13.0
   */
  public List<?> getList(int index) {
    return (List<?>) get(index);
  }

  /**
   * Retrieves the value of the column at the given index as a Java Map
   *
   * @param index The index of target column
   * @return A Java Map
   * @since 1.13.0
   */
  public Map<?, ?> getMap(int index) {
    return (Map<?, ?>) get(index);
  }

  /**
   * Generates a string value to represent the content of this row.
   *
   * @return A String value representing the content of this row
   * @since 0.9.0
   */
  @Override
  public String toString() {
    return this.scalaRow.toString();
  }

  /**
   * Creates a {@code Row} based on the given values.
   *
   * @param values Row elements.
   * @return A {@code Row} object.
   * @since 0.9.0
   */
  public static Row create(Object... values) {
    return new Row(values);
  }

  com.snowflake.snowpark.Row getScalaRow() {
    return this.scalaRow;
  }
}
